.github/workflows/ci.yml

```diff
@@ -379,12 +381,63 @@ jobs:
           composer config allow-plugins.silverstripe/recipe-plugin true
           composer config allow-plugins.silverstripe/vendor-plugin true
 
-          # Useful to see generated composer.json when diagnosing new bugs
-          cat composer.json
-
           # matrix.composer_args sometimes includes `--prefer-lowest` which is only supported by `composer update`, not `composer install`
           # Modules do not have composer.lock files, so `composer update` is the same speed as `composer install`
-          composer update --no-interaction --no-progress ${{ matrix.composer_args }}
+          # `|| :` prevents an exit code on a failed composer update from halting the workflow
+          composer update --no-interaction --no-progress ${{ matrix.composer_args }} 2> __update_attempt.txt || :
+
+          if ! [[ $(cat __update_attempt.txt) =~ Problem ]]; then
+            # Succesfully ran composer update and installed everything
+            rm __update_attempt.txt
+            # Useful to see generated composer.json when diagnosing new bugs
+            cat composer.json
+          elif ! [[ $(cat __update_attempt.txt) =~ 'Root composer.json requires silverstripe/installer' ]] || \
+            [[ $INPUTS_COMPOSER_REQUIRE_EXTRA =~ silverstripe/installer ]] || \
+            [[ $MATRIX_COMPOSER_REQUIRE_EXTRA =~ silverstripe/installer ]] || \
+            ! [[ "${{ matrix.installer_version }}" =~ ^([1-9])\.([0-9]+)\.x\-dev$ ]]
+          then
+            # Failed to run composer update and will not attempt requiring different versions of silverstripe/installer
+            cat composer.json
+            cat __update_attempt.txt
+            exit 2
+          else
+            # gha-generate-matrix will not always provide a compatibile version of silverstripe/installer due to
+            # limitations of knowing what versions of silverstripe/installer the pushed code is compatible with
+            # In this scenario we will usually end up with the latest minor x.dev version of silverstripe installer
+            # If it did initially work on composer update, try some earlier versions of silverstripe/installer
+            rm __update_attempt.txt
+            echo "Could not run composer update with version of silverstripe/installer, attempting earlier versions"
+            MAJOR=${BASH_REMATCH[1]}
+            MINOR=${BASH_REMATCH[2]}
+            SUCCESSFULLY_REQUIRED='false'
+            for MI in $(( MINOR - 1 )) $(( MINOR - 2 )); do
+              if [[ $SUCCESSFULLY_REQUIRED == 'false' ]]; then
+                # Test if $MI is a negative number
+                if [[ $MI =~ - ]]; then
```

Comment:
Can we test `< 0` instead for readability?
If it's not that simple then this is fine.

Reply:
Not really,  bash can be unbelievably difficult to do really simple things with

```
MYVAR=-1
if [[ (( $MYVAR < 2 )) ]]; then echo 't'; else echo 'f'; fi
if [[ (( $MYVAR < 0 )) ]]; then echo 't'; else echo 'f'; fi

t
f
```

amazing right?

Reply:
Amazing lol. This is fine then.